docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,192 @@
+than to reimplement the Tendermint socket protocol.
+
+Moreover, gRPC addresses all the above issues out-of-the-box, and requires
+(much) less custom code for both the server (i.e., the application) and the
+client. The project is well-funded and widely-used, which makes it a safe bet
+for a dependency.
+
+## Decision
+
+There is a set of related alternatives to consider:
+
+- Question 1: Designate a single IPC standard for out-of-process applications?
+
+  Claim: We should converge on one (and only one) IPC option for out-of-process
+  applications. We should choose an option that, after a suitable period of
+  deprecation, can address most or all the highest-impact uses of Tendermint.
+  Maintaining multiple options increases the surface area for bugs and
+  vulnerabilities, and we should not have multiple options for basic interfaces
+  without a clear and well-documented reason.
+
+- Question 2a: Standardize on gRPC and deprecate/remove the socket protocol?
+
+  Claim: Maintaining and improving a custom RPC protocol is a substantial
+  project and not directly relevant to the requirements of consensus. We would
+  be better served by depending on a well-maintained open-source library like
+  gRPC.
+
+- Question 2b: Improve the socket protocol and deprecate/remove gRPC?
+
+  Claim: If we find meaningful advantages to maintaining our own custom RPC
+  protocol in Tendermint, we should treat it as a first-class project within
+  the core and invest in making it good enough that we do not require other
+  options.
+
+**One important consideration** when discussing these questions is that _any
+outcome which includes keeping the socket protocol will have eventual migration
+impacts for clients_ regardless. To fix the limitations of the socket protocol
+as it is currently designed will require making _breaking changes_ to the
+protocol.  So, while we may put off a migration cost for clients by retaining
+the socket protocol in the short term, we will eventually have to pay those
+costs to fix the problems in its current design.
+
+## Detailed Design
+
+1. If we choose to standardize on gRPC, the main work in in Tendermint core
+   will be removing and cleaning up the code for the socket client and server.
+
+   Besides the code cleanup, we will also need to clearly document a
+   deprecation schedule, and invest time in making the migration easier for
+   applications currently using the socket protocol.
+
+   > **Point for discussion:** Migrating from the socket protocol to gRPC
+   > should mostly be a plumbing change, as long as we do it during a release
+   > in which we are not making other breaking changes to ABCI. However, the
+   > effort may be more or less depending on how gRPC integration works in the
+   > application's implementation language, and would have to be sure networks
+   > have plenty of time not only to make the change but to verify that it
+   > preserves the function of the network.
```

Comment:
My operating assumption was that a clear deprecation notice alongside the 0.36 release would give users an entire release cycle until v0.37, which I believe to be sufficient, to trial the migration and to provide feedback if unexpected hiccups were to arise. 

In the case of https://github.com/informalsystems/tendermint-rs/tree/master/abci or https://github.com/penumbra-zone/tower-abci which I believe are the predominant known implementations, we may be able to assist in the migration

Reply:
In terms of communicating the schedule, I think "one release" is probably fine. The larger question I'm concerned with is what's the impact: How many applications does it affect, how big a change is it for those specific apps, how much lead time will developers and operators need to test the change, and so forth.

Mainly I see this as a visibility issue—we can write lots of documentation and give lead time, but developers who don't see it or don't realize they need to take it seriously could still be caught out.

Reply:
There's no rush about merging. I plan to keep this open till we've circulated it more widely anyway.

Reply:
I agree that we need the opinion of the `tendermint-rs` guys, but I don't think this should prevent this PR to be merged. I suggest opening an Issue to track this discussion about Rust support.